### ISSUE TITLE
There are view generation errors

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#4612]
Locally (using data from demo.dev) I have noticed that somewhere in medic-client (and perhaps elsewhere too, I haven't checked), some views are throwing errors in CouchDB.

[Up with this we will not put!](https://englishlanguagethoughts.com/2016/06/03/up-with-this-i-will-not-put/).

Someone should:
 - Get a bunch of reasonable data on their system
 - Force view regeneration by slightly altering a view in each ddoc
 - Watch the log for errors
 - If there are errors, divide and conquer by removing views until you narrow down which ones are causing the problems, and fix them

**Comment by SCdF** [comment]
lol github no.

Edit incoming.

**Comment by SCdF** [comment]
Look again, notification squad, the title of this ticket has been updated to "There are view generation errors" and there is now a description

**Comment by vimemo** [comment]
Unable to reproduce. Have replicated data from demo.dev and other instances. The only errors in my local logs have to do with compilation errors that I have provoked while updating/testing views. @SCdF  , do you have any suggestions? 
I am using couch 2 in my local env. Thank you.

**Comment by SCdF** [comment]
@vimago that's weird that I got it and you didn't. What version of the code were you running? demo.dev will not be on 3.0

**Comment by vimemo** [comment]
hi,  @SCdF. 
I tested using latest from master but did go through the logs in demo.dev. There is no errors related to views there. demo uses couch 1.7. Should we let this go and assume that demo will soon go to couch2 and latest code?
Otherwise, I can do a bit more exploration with the exact code and db version but we don't even have those errors in that env. what do you think?

**Comment by garethbowen** [comment]
I just caught this one on beta-old.dev:

```
[Wed, 04 Jul 2018 02:52:51 GMT] [info] [<0.12097.47>] Starting index update for db: medic idx: _design/medic
[Wed, 04 Jul 2018 02:52:51 GMT] [info] [<0.30204.4>] OS Process #Port<0.33759> Log :: function raised exception (new TypeError("task.messages is undefined", "undefined", 4)) with doc._id a1d31848b02e9c38d353df34122fdc24
[Wed, 04 Jul 2018 02:52:51 GMT] [info] [<0.12097.47>] Index update finished for db: medic idx: _design/medic
```

The view is `medic/tasks_messages` in `2.16.x` (it's been moved/renamed in `master`).

**Comment by garethbowen** [comment]
@vimemo The `medic-sms/tasks_messages` view code is the same in `master` so it could be the cause of this issue. To check make sure you force a view regeneration on the `medic-sms` ddoc.

Either way I think we should add a null check in the view map function to avoid this happening in future. Even bad data (which I believe this was) shouldn't take out all the view generation...

**Comment by SCdF** [comment]
Merged. Moving to ready, I don't think this needs to be AT'd.

**Comment by garethbowen** [comment]
Reopening this to get it backported to 2.18.0. @dianabarsan Are you able to take this on, please?

**Comment by dianabarsan** [comment]
I have backported the view map fixes to `2.18`. 

To AT, an example scenario would be: using current `standard` config, register a new pregnancy via SMS, visit the report page to make sure the Automated reply exists and there are a bunch of ANC Reminders Outgoing Messages.

The automated reply status should be `pending`.
Now, visit `<your_instance>/api/sms` and check the app to verify that the Automated Reply received a `forwarded to gateway` status. (Note. `/api/sms` processes only 25 messages at a time, so if your queue of outgoing messages is larger, you might need to call it multiple times).

**Comment by ngaruko** [comment]
LGTM. Tested on `standardard.dev` v2.18 with [this report](https://standard.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5f8bcedcbb7cdab700900d443ecda1f0)
